src/Npgsql/Util/TimeoutCancellationTokenSourceWrapper.cs

```diff
@@ -15,37 +17,114 @@ class TimeoutCancellationTokenSourceWrapper : IDisposable
+                }
             }
 
+            _linkedToken = cancellationToken;
+#if DEBUG
+            _isRunning = true;
+#endif
             return _timeoutCts.Token;
         }
 
         /// <summary>
-        /// Reset the timeout on the wrapped <see cref="CancellationTokenSource"/>
+        /// Restart the timeout on the wrapped <see cref="CancellationTokenSource"/> without reinitializing it,
+        /// even if <see cref="IsCancellationRequested"/> is already set to <see langword="true"/>
         /// </summary>
-        public void Restart() => _timeoutCts.CancelAfter(Timeout);
+        public void RestartTimeoutWithoutReset() => _timeoutCts.CancelAfter(Timeout);
+
+        /// <summary>
+        /// Reset the wrapper to contain a unstarted and uncancelled <see cref="CancellationTokenSource"/>
+        /// in order make sure the next call to <see cref="Start"/> will not invalidate
+        /// the cancellation token.
+        /// </summary>
+        /// <param name="cancellationToken">An optional cancellation token that will be linked with the
+        /// contained <see cref="CancellationTokenSource"/></param>
+        /// <returns>The <see cref="CancellationToken"/> from the wrapped <see cref="CancellationTokenSource"/></returns>
+        public CancellationToken Reset(CancellationToken cancellationToken = default)
+        {
+            if ((_linkedToken.CanBeCanceled || cancellationToken.CanBeCanceled) && _linkedToken != cancellationToken)
+            {
+                _timeoutCts.Dispose();
+                _timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            }
+            else
+            {
+                _timeoutCts.CancelAfter(InfiniteTimeSpan);
+                if (_timeoutCts.IsCancellationRequested)
+                {
+                    _timeoutCts.Dispose();
+                    _timeoutCts = new CancellationTokenSource();
+                }
+            }
+
+            _linkedToken = cancellationToken;
+#if DEBUG
+            _isRunning = false;
+#endif
+            return _timeoutCts.Token;
+        }
 
         /// <summary>
         /// Set the timeout on the wrapped <see cref="CancellationTokenSource"/>
         /// to <see cref="System.Threading.Timeout.InfiniteTimeSpan"/>
         /// </summary>
-        public void Stop() => _timeoutCts.CancelAfter(System.Threading.Timeout.InfiniteTimeSpan);
+        /// <remarks>
+        /// <see cref="IsCancellationRequested"/> can still arrive at a state
+        /// where it's value is <see langword="true"/> if the <see cref="CancellationToken"/>
+        /// passed to <see cref="Start"/> gets a cancellation request.
+        /// If this is the case it will be resolved upon the next call to <see cref="Start"/>
+        /// or <see cref="Reset"/>
+        /// </remarks>
+        public void Stop()
+        {
+#if DEBUG
+            Debug.Assert(_isRunning);
+#endif
```

Comment:
That would only work if I'd make the `_isRunning` variable available unconditionally.
The compiler will not translate the method to IL but it can't automatically hide the method argument from syntax checking so it would be reported as CS0103.

Reply:
OK, doesn't really matter...  We can also always have _isRunning but only read/write under DEBUG (and suppressed the "unused field" warning), but it really doesn't matter much.